scripts/gen_mmu.py

```diff
@@ -0,0 +1,425 @@
+
+#     union x86_mmu_pde_pt {
+# 	u32_t  value;
+# 	struct {
+# 		u32_t p:1;
+# 		u32_t rw:1;
+# 		u32_t us:1;
+# 		u32_t pwt:1;
+# 		u32_t pcd:1;
+# 		u32_t a:1;
+# 		u32_t ignored1:1;
+# 		u32_t ps:1;
+# 		u32_t ignored2:4;
+# 		u32_t page_table:20;
+# 	};
+# };
+
+def page_directory_create_binary_file():
+    global output_buffer
+    global output_offset
+    for pde in range(1024):
+        binary_value = 0 # the page directory entry is not valid
+
+        # if i have a valid entry to populate
+        if pde in list_of_pde.keys():
+            value = list_of_pde[pde]
+            present = 1 << 0;
+            read_write = ( ( value.permissions >> 1) & 0x1) << 1;
+            user_mode = ( ( value.permissions >> 2) & 0x1) << 2;
+            pwt = 0 << 3;
+            pcd = 0 << 4;
+            a = 0 << 5; # this is a read only field
+            ps = 0 << 7; # this is a read only field
+            page_table = address_of_page_table(value.pde_index) << 12;
+            binary_value = (present | read_write | user_mode | pwt | pcd | a | ps | page_table)
+
+        struct.pack_into(write_4byte_bin,output_buffer, output_offset, binary_value)
+        output_offset += struct.calcsize(write_4byte_bin)
+
+
+# array of 1024 only the required memory regions pointing to a page table.
+def page_directory_create_c_file():
+    pde_string = ''
+    global pd_complete
+    for key, value in list_of_pde.items():
+        temp_string = template_enabled_pde.replace("PDE_NUMBER",str(value.pde_index))
+        temp_string = temp_string.replace("PERMISSIONS",
+                                          str((hex((value.permissions >> 1) & 0x1))))
+        temp_string = temp_string.replace("SUPERVISOR",
+                                          str((hex((value.permissions >> 2) &  0x1))))
+        temp_string = temp_string.replace("ADDRESS_OF_PAGE_TABLE","&page_table_"+str(key))
+        pde_string += temp_string
+
+    pd_complete = template_pd_complete.replace("REPACEMENT_HOOK",pde_string)
+
+    output_file.write(pd_complete)
+
+# union x86_mmu_pte {
+# 	u32_t  value;
+# 	struct {
+# 		u32_t p:1;
+# 		u32_t rw:1;
+# 		u32_t us:1;
+# 		u32_t pwt:1;
+# 		u32_t pcd:1;
+# 		u32_t a:1;
+# 		u32_t d:1;
+# 		u32_t pat:1;
+# 		u32_t g:1;
+# 		u32_t alloc:1;
+# 		u32_t custom:2;
+# 		u32_t page:20;
+# 	};
+# };
+
+def page_table_create_binary_file():
+    global output_buffer
+    global output_offset
+    for key, value in list_of_pde.items():
+        for pte in range(1024):
+            binary_value = 0 # the page directory entry is not valid
+
+            # print(value.pte_valid_addr_start,value.pte_valid_addr_end )
+            # if i have a valid entry to populate
+            if (pte >= value.pte_valid_addr_start) and (pte <= value.pte_valid_addr_end):
+                present = 1 << 0;
+                read_write = ( ( value.permissions >> 1) & 0x1) << 1;
+                user_mode = ( ( value.permissions >> 2) & 0x1) << 2;
+                pwt = 0 << 3;
```

Comment:
kept it as a place holder.
Whenever there is a need to set that bit , I just have to replace "0" with "1".
Later on I need not worry about bit positions.